pilot/pkg/dataplane/interfaces.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dataplane
+
+import (
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// Dataplane represents the interfaces to be implemented by code that generates xDS responses
+type Dataplane interface {
+	// BuildListeners returns the list of listeners for the given proxy. This is the LDS output
```

Comment:
I'm not sure I like this: part of LDSv2 we want to have far more granularity, and having an interface like this doesn't seem to help. 

Who would implement it ? Who would use it ? It doesn't really fit on either end - the grpc interfaces are push based, and need to react to events by building only the listener that changed ( typically a new service shows up). We certainly don't want to build all listners again and again.

Reply:
AFAIK, all listeners change when a new service shows up.

Reply:
We don't have incremental xDS for envoy. You send all listeners, all clusters, and all routes. Incremental xDS is coming, but its going to take a while.

Also, this is just the first cut to organize the existing code based on the existing usage patterns. There is a separate issue on how to notify this dataplane about service registry changes, and how the data plane will notify the xds server about changes to its configuration. those are two interfaces that can be added..